docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

```diff
@@ -0,0 +1,268 @@
+   function is called by the SDK's enclave-side runtime prior to the initialization
+   of the module.
+
+- `oe_result_t oe_debug_notify_module_unloaded(oe_debug_module_t* module)`
+
+   This function is called by the SDK runtimes (host-side or enclave-side), after a
+   module has been unloaded.
+
+Debugger Handling of Module Load/Unload Notifications
+
+- Linux
+  Similar to enclave load/unload, the debugger sets internal breakpoints in
+  `oe_debug_notify_module_loaded` and `oe_debug_notify_module_unloaded`
+  functions. When the SDK runtimes call these functions, the breakpoint is
+  triggered and the debugger get control. The debugger reads the module information
+  from the `oe_debug_module_t` structure either loads or unloads the debugging
+  information for the module.
+
+- Windows
+  - Host Side
+  
+    The SDK host-side runtime uses the `RaiseException` Windows API to
+	communicate with the debugger.
+	`oe_debug_notify_module_loaded` raises the following exception:
+	```c
+ 
+        ULONG_PTR args[1] = {(ULONG_PTR)module};
+		RaiseException(
+		  OE_DEBUGRT_MODULE_LOADED_EVENT,
+		  0, // dwFlags
+		  1, // number of args
+		  args);
+		  
+    ```
+	
+     whereas `oe_debug_notify_enclave_terminated` raises:
+     ```c
+   
+        ULONG_PTR args[1] = {(ULONG_PTR)module};
+        RaiseException(
+          OE_DEBUGRT_MODULE_UNLOADED_EVENT,
+          0, // dwFlags
+          1, // number of args
+          args);
+		  
+	```
+	
+  - Enclave side
+    
+	The enclave side runtime cannot call Windows API functions. Instead, it
+	relies on `int 3` exception to communicate with the debugger. The `int 3`
+	instruction is preceeded by special byte sequence `0x0edeb0gca11bac00` to
+	allow the debugger to distinguish between these special `int 3` exceptions
+	and any other `int 3` exceptions that may occur.
+	
+	`oe_debug_notify_module_loaded` uses the following instruction pattern:
+	```asm
+
+       mov module, %rax
+	   mov $1, %rbx     // 1 => ModuleLoaded
+	   jmp do_int3
+	   .bytes 0x0edeb0gca11bac00
+	do_int3:
+	   int $3
+	   
+	```
+	whereas `oe_debug_notify_module_unloaded` uses the following instruction
+	pattern:
+	```asm
+	   
+	   mov module, %rax
+	   mov $2, %rbx     // 2 => ModuleUnloaded
+	   jmp do_int3
+	   .bytes 0x0edeb0gca11bac00
+	do_int3:
+	   int $3
+	   
+	```
+
+    When the debugger receives and `int 3` exception, it first checks the bytes
+	immediately preceeding the program counter. If the bytes match the special
+	sequence, then the debugger reads `rax` register to obtain the pointer to
+	`oe_debug_module_t` structure, and `rbx` register to determine whether
+	the module was loaded or unloaded.
+
+
+The Debugger Contract version is incremented to `2` to reflect the addition of
+the notification functions.
+
+The host-side runtime will set [oe_debugger_contract_version](https://github.com/openenclave/openenclave/blob/6f94547dc920d70eb50ec8ca95ab65a17597364b/include/openenclave/internal/debugrt/host.h#L131)
```

Comment:
Note: I don't believe the Windows debuggers will currently do this, though this could definitely be a future behavior.